Automattic/wp-calypso
# Move the "anchor fm podcast id" state from redux to "location state"

#### Parent Branch

This merges into the branch of PR #47600 ; for Anchor.fm flavored gutenboarding.

#### Changes proposed in this Pull Request

* Move the "anchor fm podcast id" state from redux to "location state" provided by redux router
- Fixes F5 problem, where starting the anchor flow, then pressing F5 on a step moves you to the non-anchor flow

#### Testing instructions

* [Anchor] F5 after signup (delayed) - Reproduce existing bug
  - In the use-step-navigation file, console.log the value of `isAnchorFmSignup`
  - http://calypso.localhost:3000/new?anchor_podcast=2&signup
  - Insert new email, password, click create account
  - Fill out a site name, click continue
  - Now on choose a design page. Open console, note that `isAnchorFmSignup` is true
  - F5.  Same page loads, but in console, `isAnchorFmSignup` is now false.

That's how to reproduce the bug before the PR.  After the PR, the same instructions will work, except you now log `anchorFmPodcastId`, and when you press F5, it doesn't go away.

* [Non-Anchor] Regular gutenboarding should work the same
  - http://calypso.localhost:3000/new continues to work

#### Explanation Comment from 47600

For regular gutenboarding, pressing F5 does not remove state. For example, you might choose font #3, move to another step, press F5, and the page still remembers that you have selected that font, even having reloading. The redux state is persisted.

For this branch, pressing F5 wipes out the anchor podcast id sometimes. That's because:

  - We migrate the podcast id from the query parameter to redux
  - Clicking next/previous steps removes the query parameter from the url
  - Pressing F5 causes the podcast id to be re-migrated from the query parameter to redux, so even though the redux key is preserved, this re-migration effect hook wipes it out.

I have a prospective solution in a PR that merges into this branch here: #47941. It removes the anchor podcast ID from redux and stores it in "location state" handled by react-router, and adds that state to be passed on when clicking the next/previous buttons. It fixes the F5 problem in my testing.

Possibly there's a way to solve the F5 problem while also storing the value in redux and continuing to migrate it from the query param, but I didn't find one. Open to alternate suggestions or reviews.
